base/channels.jl

```diff
@@ -316,17 +317,36 @@ function put!(c::Channel{T}, v) where T
     return isbuffered(c) ? put_buffered(c, v) : put_unbuffered(c, v)
 end
 
+# Atomically update channel length, *assuming* we hold the channel lock.
+macro _inc_channel_length(c, inc)
+    quote
+        c = $(esc(c))
+        # We hold the channel lock so it's safe to non-atomically read and
+        # increment c.length
+        newlen = c.length + $(esc(inc))
+        # Atomically store c.length to prevent data races with other threads
+        # reading the length outside the lock.
+        @atomic :monotonic c.length = newlen
```

Comment:
By the way, if you're wondering "what? why do we need a macro for this?". Well we don't of course, but it gives a single location to clearly comment on the correct way to update the length field.

Reply:
why isn't it just a function?

Reply:
An extremely good question :sweat_smile: Fixed now.